proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -67,101 +68,597 @@ store values that may only be determined at runtime.
+vector in memory along with the format of each element.
+
+As with `dx::linalg::MatrixRef`, two versions are provided - one for vectors
+stored in `ByteAddressBuffer` and another for `RWByteAddressBuffer`. A base
+class, `dx::linalg::VectorRefImpl`, covers both of them.
+
+Example usage:
+
+```c++
+ByteAddressBuffer ROBuffer;
+RWByteAddressBuffer RWBuffer;
+
+void Example() {
+  using namespace dx::linalg;
+
+  VectorRef<DATA_TYPE_FLOAT16> VectorA = {ROBuffer, /*offset=*/128};
+  VectorRef<DATA_TYPE_FLOAT32> VectorB = {ROBuffer, /*offset=*/128};
+  RWVectorRef<DATA_TYPE_SINT16> VectorC = {RWBuffer, /*offset=*/64};
+}
+```
+
+Template parameter:
+
+- `DT` - the data type of each element stored in the buffer
+
+Members:
+
+- `Buffer` - the buffer that the vector is stored in
+  - For `VectorRef` this is a `ByteAddressBuffer`
+  - For `RWVectorRef` this is a `RWByteAddresssBuffer`
+- `StartOffset` - the offset, in bytes, from the beginning of the buffer to
+  where the vector is located.
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename BufferTy, DataType DT> struct VectorRefImpl {
+  BufferTy Buffer;
+  uint StartOffset;
+};
+
+template <DataType DT> using VectorRef = VectorRefImpl<ByteAddressBuffer, DT>;
+
+template <DataType DT>
+using RWVectorRef = VectorRefImpl<RWByteAddressBuffer, DT>;
+
+} // namespace linalg
+} // namespace dx
+
+```
+
+### struct Vector
+
+> NOTE: it's possible that one resolution of [441] may remove the need for this
+> type entirely.
+
+The `dx::linalg::Vector` struct is a wrapper around `vector`, adding an
+interpretation value that controls how the data in the vector should be
+interpreted. Although the struct can be used directly, it is likely more
+ergonomic to use the `dx::linalg::InterpretedVector` function that's also
+described here.
+
+Example usage:
+
+```c++
+ByteAddressBuffer Buffer;
+void Example() {
+  using namespace dx::linalg;
+
+  MatrixRef<DATA_TYPE_FLOAT16, 128, 128, MATRIX_LAYOUT_MUL_OPTIMAL, true>
+      Matrix = {Buffer, 0, 0};
+
+  vector<float, 128> V = 0;
+  vector<float, 128> Result =
+      Mul<float>(Matrix, InterpretedVector<DATA_TYPE_E4M3>(V));
+
+  // alternative:
+  Vector<float, 128, DATA_TYPE_E4M3> IV = {V};
+  vector<float, 128> Result2 = Mul<float>(Matrix, IV);
+}
+```
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename T, int N, DataType DT> struct Vector {
+  vector<T, N> Data;
+};
+
+template <DataType DT, typename T, int N>
+Vector<T, N, DT> InterpretedVector(vector<T, N> Vec) {
+  Vector<T, N, DT> IV = {Vec};
+  return IV;
+}
+
+} // namespace linalg
+} // namespace dx
+```
+
+[441]: https://github.com/microsoft/hlsl-specs/issues/441
+
+### Function: Mul
+
+The `dx::linalg::Mul` function performs a matrix-vector multiplication. The
+matrix is stored in memory, while the vector comes from a variable.
+
+> TODO: add an example for packed types, and make sure they work correctly
```

Comment:
I assume this TODO is intended to be completed before this spec is checked in?

Reply:
I think this PR can be completed and the packed example added in a subsequent one.